### PR TITLE
Distributor: Return 529 for ingestion rate limit when serviceOverloadErrorEnabled

### PR DESCRIPTION
**Pull request by ying-jeanne**
This has been overlooked. The flag was introduced in this [PR](https://github.com/grafana/mimir/pull/5752). We also need to implement a similar approach for the ingestion rate limit to prevent data loss when rate limiting occurs.

Comment is equally added into doc, since with load test, I have observed that returning 529 instead of 429 could result in enhancing thundering herd. See test result

Metrics data accumulates for 5 minutes with 1000 agents, after which the network is enabled to transmit data to Mimir. This results in a 429 error code, impacting the request rate on the gateway:
Screenshot 2023-11-10 at 00 28 53

with 529 error code, request rate on gateway:
Screenshot 2023-11-10 at 00 29 05

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [x] Tests updated
- na Documentation added
- na `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by pstibrany — COMMENTED** [review]
I think this change deserves a changelog entry.

**Review by aknuds1 — COMMENTED** [review]
Agree w/ Peter there should be a changelog entry. Should this be backported to v2.10 since it fixes #5752, which is in that release?

**Comment by pstibrany** [comment]
> Should this be backported to v2.10 since it fixes #5752, which is in that release?

This is experimental feature, even we don't use it anywhere internally yet.

**Comment by ying-jeanne** [comment]
@aknuds1 @pstibrany It is important to give it whatever its deserved. PR updated, could you give it another look. thanks

File: pkg/util/validation/limits.go [hunk 1]
```diff
@@ -205,7 +205,7 @@ func (l *Limits) RegisterFlags(f *flag.FlagSet) {
...
 	f.BoolVar(&l.EnforceMetadataMetricName, "validation.enforce-metadata-metric-name", true, "Enforce every metadata has a metric name.")
-	f.BoolVar(&l.ServiceOverloadStatusCodeOnRateLimitEnabled, "distributor.service-overload-status-code-on-rate-limit-enabled", false, "If enabled, rate limit errors will be reported to the client with HTTP status code 529 (Service is overloaded). If disabled, status code 429 (Too Many Requests) is used.")
+	f.BoolVar(&l.ServiceOverloadStatusCodeOnRateLimitEnabled, "distributor.service-overload-status-code-on-rate-limit-enabled", false, "If enabled, rate limit errors will be reported to the client with HTTP status code 529 (Service is overloaded). If disabled, status code 429 (Too Many Requests) is used. Enabling distributor.retry-after-header.enabled before utilizing this option is strongly recommended as it helps prevent premature request retries by the client.")
```
Comment (pstibrany) [inline pkg/util/validation/limits.go:208]:
```suggestion
	f.BoolVar(&l.ServiceOverloadStatusCodeOnRateLimitEnabled, "distributor.service-overload-status-code-on-rate-limit-enabled", false, "If enabled, rate limit errors will be reported to the client with HTTP status code 529 (Service is overloaded). If disabled, status code 429 (Too Many Requests) is used. Enabling -distributor.retry-after-header.enabled before utilizing this option is strongly recommended as it helps prevent premature request retries by the client.")
```

**Review by pstibrany — APPROVED** [review]

**Pull request merged**